fuel-types/src/canonical.rs

```diff
@@ -0,0 +1,619 @@
+pub trait Deserialize: Sized {
+    /// !INTERNAL USAGE ONLY!
+    /// Array of bytes that are now aligned by themselves.
+    #[doc(hidden)]
+    const UNALIGNED_BYTES: bool = false;
+
+    /// Decodes `Self` from the `buffer`.
+    ///
+    /// It is better to not implement this function directly, instead implement
+    /// `decode_static` and `decode_dynamic`.
+    fn decode<I: Input + ?Sized>(buffer: &mut I) -> Result<Self, Error> {
+        let mut object = Self::decode_static(buffer)?;
+        object.decode_dynamic(buffer)?;
+        Ok(object)
+    }
+
+    /// Decodes static part of `Self` from the `buffer`.
+    fn decode_static<I: Input + ?Sized>(buffer: &mut I) -> Result<Self, Error>;
+
+    /// Decodes dynamic part of the information from the `buffer` to fill `Self`.
+    ///
+    /// # Note: It is empty for primitives. But it can be helpful for containers to fill elements.
+    fn decode_dynamic<I: Input + ?Sized>(
+        &mut self,
+        _buffer: &mut I,
+    ) -> Result<(), Error> {
+        Ok(())
+    }
+
+    /// Helper method for deserializing `Self` from bytes.
+    fn from_bytes(mut buffer: &[u8]) -> Result<Self, Error> {
+        Self::decode(&mut buffer)
+    }
+}
+
+/// The data of each field should be 64 bits aligned.
+pub const ALIGN: usize = 8;
+
+/// Returns the number of bytes to fill aligned
+const fn alignment_bytes(len: usize) -> usize {
+    (ALIGN - (len % ALIGN)) % ALIGN
+}
+
+/// Size after alignment
+pub const fn aligned_size(len: usize) -> usize {
+    len + alignment_bytes(len)
+}
+
+macro_rules! impl_for_fuel_types {
+    ($t:ident) => {
+        impl SerializedSize for $t {
+            // Fuel-types are transparent single-field structs, so the size matches
+            const SIZE_STATIC: usize = aligned_size(::core::mem::size_of::<$t>());
+        }
+
+        impl Serialize for $t {
+            const SIZE_NO_DYNAMIC: bool = true;
+
+            #[inline(always)]
+            fn encode_static<O: Output + ?Sized>(
+                &self,
+                buffer: &mut O,
+            ) -> Result<(), Error> {
+                for _ in 0..alignment_bytes(self.as_ref().len()) {
+                    buffer.push_byte(0)?;
+                }
+                buffer.write(self.as_ref())?;
+                Ok(())
+            }
+        }
+
+        impl Deserialize for $t {
+            fn decode_static<I: Input + ?Sized>(buffer: &mut I) -> Result<Self, Error> {
+                let mut asset = $t::zeroed();
+                buffer.skip(alignment_bytes(asset.as_ref().len()))?;
+                buffer.read(asset.as_mut())?;
+                Ok(asset)
+            }
+        }
+    };
+}
+
+impl_for_fuel_types!(Address);
+impl_for_fuel_types!(AssetId);
+impl_for_fuel_types!(ContractId);
+impl_for_fuel_types!(Bytes4);
+impl_for_fuel_types!(Bytes8);
+impl_for_fuel_types!(Bytes20);
+impl_for_fuel_types!(Bytes32);
+impl_for_fuel_types!(MessageId);
+impl_for_fuel_types!(Salt);
+impl_for_fuel_types!(Nonce);
```

Comment:
You don't need to implement it manually. You can derive it because these types are defined in the same crate.

During the previous implementation, I defined traits in the `fuel-tx` which is why I should implemented them manually.

Reply:
Migrated to derived impl in 34e6a048